From 82cd321875ae207aead8b065e7e4fb5ee416a395 Mon Sep 17 00:00:00 2001
From: Tanish2002 <tanishkhare@gmail.com>
Date: Wed, 27 Jan 2021 16:46:35 +0530
Subject: [PATCH 15/16] Add iso14755

---
 config.def.h |  1 +
 st.1         |  4 ++++
 st.c         | 25 +++++++++++++++++++++++++
 st.h         |  2 ++
 4 files changed, 32 insertions(+)

diff --git a/config.def.h b/config.def.h
index bd38b7d..a9a29e6 100644
--- a/config.def.h
+++ b/config.def.h
@@ -268,6 +268,7 @@ static Shortcut shortcuts[] = {
 	{ TERMMOD,              XK_Y,           selpaste,       {.i =  0} },
 	{ ShiftMask,            XK_Insert,      selpaste,       {.i =  0} },
 	{ TERMMOD,              XK_Num_Lock,    numlock,        {.i =  0} },
+	{ TERMMOD,              XK_I,           iso14755,       {.i =  0} },
 	{ ShiftMask,            XK_Page_Up,     kscrollup,      {.i = -1} },
 	{ ShiftMask,            XK_Page_Down,   kscrolldown,    {.i = -1} },
 	{ TERMMOD,              XK_Return,      newterm,        {.i =  0} },
diff --git a/st.1 b/st.1
index 39120b4..4a98626 100644
--- a/st.1
+++ b/st.1
@@ -159,6 +159,10 @@ Copy the selected text to the clipboard selection.
 .TP
 .B Ctrl-Shift-v
 Paste from the clipboard selection.
+.TP
+.B Ctrl-Shift-i
+Launch dmenu to enter a unicode codepoint and send the corresponding glyph
+to st.
 .SH CUSTOMIZATION
 .B st
 can be customized by creating a custom config.h and (re)compiling the source
diff --git a/st.c b/st.c
index a676653..16cffa4 100644
--- a/st.c
+++ b/st.c
@@ -48,6 +48,9 @@
 				term.scr + HISTSIZE + 1) % HISTSIZE] : \
 				term.line[(y) - term.scr])
 
+/* constants */
+#define ISO14755CMD		"dmenu -w \"$WINDOWID\" -p codepoint: </dev/null"
+
 enum term_mode {
 	MODE_WRAP        = 1 << 0,
 	MODE_INSERT      = 1 << 1,
@@ -2182,6 +2185,28 @@ tprinter(char *s, size_t len)
 	}
 }
 
+void
+iso14755(const Arg *arg)
+{
+	FILE *p;
+	char *us, *e, codepoint[9], uc[UTF_SIZ];
+	unsigned long utf32;
+
+	if (!(p = popen(ISO14755CMD, "r")))
+		return;
+
+	us = fgets(codepoint, sizeof(codepoint), p);
+	pclose(p);
+
+	if (!us || *us == '\0' || *us == '-' || strlen(us) > 7)
+		return;
+	if ((utf32 = strtoul(us, &e, 16)) == ULONG_MAX ||
+	    (*e != '\n' && *e != '\0'))
+		return;
+
+	ttywrite(uc, utf8encode(utf32, uc), 1);
+}
+
 void
 toggleprinter(const Arg *arg)
 {
diff --git a/st.h b/st.h
index de307f6..93287f3 100644
--- a/st.h
+++ b/st.h
@@ -106,6 +106,7 @@ void kscrolldown(const Arg *);
 void kscrollup(const Arg *);
 void newterm(const Arg *);
 void externalpipe(const Arg *);
+void iso14755(const Arg *);
 void printscreen(const Arg *);
 void printsel(const Arg *);
 void sendbreak(const Arg *);
@@ -156,4 +157,5 @@ extern char *termname;
 extern unsigned int tabspaces;
 extern unsigned int defaultfg;
 extern unsigned int defaultbg;
+extern char *iso14755_cmd;
 extern const int boxdraw, boxdraw_bold, boxdraw_braille;
-- 
2.30.0

